packages/flutter/lib/src/gestures/hit_test.dart

```diff
@@ -54,30 +61,188 @@ class HitTestEntry {
+  /// is removed by calling [PointerEvent.paintTransformToPointerEventTransform].
+  /// The inverted transformed matrix is then used to transform `position` from
+  /// the parent coordinate system to the child coordinate system
+  /// before calling the provided `hitTest` callback, which needs to implement
+  /// the actual hit testing on the child. The [Offset] provided
+  /// to the `hitTest` callback is the position exactly under the user's
+  /// finger on the touch screen (or any other pointer device) in the local
+  /// coordinate system of the child.
+  ///
+  /// If `transform` is null it will be treated as the identity transform ad
+  /// `position` is provided to the `hitTest` callback as-is. If `transform`
+  /// cannot be inverted, the `hitTest` callback is not invoked and false is
+  /// returned. Otherwise, the return value of the `hitTest` callback is
+  /// returned.
+  ///
+  /// The `position` argument may be null, which will be forwarded to the
+  /// `hitTest` callback as-is. Using null as the position can be useful if
+  /// the child speaks a different hit test protocol then the parent and the
+  /// position is not required to do the actual hit testing in that protocol.
+  ///
+  /// {@tool sample}
+  /// This method is used in [RenderBox.hitTestChildren]:
+  ///
+  /// ```dart
+  /// abstract class Foo extends RenderBox {
+  ///
+  ///   final Matrix4 _effectiveTransform = Matrix4.rotationZ(50);
+  ///
+  ///   @override
+  ///   void applyPaintTransform(RenderBox child, Matrix4 transform) {
+  ///     transform.multiply(_effectiveTransform);
+  ///   }
+  ///
+  ///   @override
+  ///   bool hitTestChildren(HitTestResult result, { Offset position }) {
+  ///     return result.withTransform(
+  ///       transform: _effectiveTransform,
+  ///       position: position,
+  ///       hitTest: (HitTestResult result, Offset position) {
+  ///         return super.hitTestChildren(result, position: position);
+  ///       },
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  /// {@end-tool}
+  ///
+  /// See also:
+  ///
+  ///  * [withPaintOffset], which can be used for `transform`s that are just
+  ///    simple matrix translations by an [Offset].
+  ///  * [withRawTransform], which takes a transform matrix that is directly
+  ///    used to transform the position without any pre-processing.
+  bool withPaintTransform({
+    @required Matrix4 transform,
+    @required Offset position,
+    @required HitTest hitTest,
+  }) {
+    assert(hitTest != null);
+    if (transform != null) {
+      transform = Matrix4.tryInvert(PointerEvent.paintTransformToPointerEventTransform(transform));
+      if (transform == null) {
+        // Objects are not visible on screen and cannot be hit-tested.
+        return false;
+      }
+      _pushTransform(transform);
+    }
+    final Offset transformedPosition = position == null ? position : PointerEvent.transformPosition(transform, position);
```

Comment:
Oh, I see. This is why this has Cartesian knowledge...

Reply:
maybe all this belongs in BoxHitTestEntry

Reply:
Yeah, the problem is that this needs to live in the HitTestResult, though, and not in an individual HitTestEntry. (There are no individual HitTestResult classes for Box or Sliver.)

My "excuse" for hard-coding the cartesian knowledge here is that the gesture library already has this cartesian knowledge baked into the events it dispatched: All PointerEvents in the gesture library have a (cartesian) Offset as position:

https://github.com/flutter/flutter/blob/8e7c7fcbd48abf2ded5ec876759375453ac18cff/packages/flutter/lib/src/gestures/events.dart#L171-L173

From that I generalized that the entire hit-testing kinda happens in the cartesian system. There is a small escape hatch for non-cartesian components: If you don't care about the transformed position, a caller can just provide null. This works pretty well in the sliver case, see https://github.com/flutter/flutter/pull/31222/files#diff-4b5ed324ca3ddab0cd1942fc98a42573R1526.

I am not entirely sure how we can move this to BoxHitTesting - we also need to collect the transformation matrix for non-box ancestors (e.g. slivers) to calculate the correct local position for pointer events.

One thing we could do is remove the Offset stuff from the gesture library and have the call-site do the position transforms if they need it. As outlined above, we'd still need to keep the transform matrix, though. I am currently not sure how "cartesian" this matrix would be.

Let me know if you have a better idea...

Reply:
Can you elaborate on "this needs to live in the HitTestResult"?

Reply:
Talked offline. We will try to untangle this by implementing separate HitTestResults for Box, Sliver, and other coordinate systems.